src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -59,12 +64,90 @@ public Task AddComponentAsync(Type componentType, ParameterView parameters, stri
+
     protected override void AttachRootComponentToBrowser(int componentId, string domElementSelector)
     {
         var attachComponentTask = _client.SendAsync("JS.AttachComponent", componentId, domElementSelector);
         _ = CaptureAsyncExceptions(attachComponentTask);
     }
 
+    protected override void UpdateRootComponents(string operationsJson)
+    {
+        var operations = JsonSerializer.Deserialize<IEnumerable<RootComponentOperation<ServerComponentMarker>>>(
+            operationsJson,
+            ServerComponentSerializationSettings.JsonSerializationOptions);
+
+        foreach (var operation in operations)
+        {
+            switch (operation.Type)
+            {
+                case RootComponentOperationType.Add:
+                    AddRootComponent(operation);
+                    break;
+                case RootComponentOperationType.Update:
+                    UpdateRootComponent(operation);
+                    break;
+                case RootComponentOperationType.Remove:
+                    RemoveRootComponent(operation);
+                    break;
+            }
+        }
+
+        return;
+
+        void AddRootComponent(RootComponentOperation<ServerComponentMarker> operation)
+        {
+            if (operation.SelectorId is not { } selectorId)
+            {
+                Log.InvalidRootComponentOperation(_logger, operation.Type, message: "Missing selector ID.");
+                return;
+            }
+
+            if (!_serverComponentDeserializer.TryDeserializeSingleComponentDescriptor(operation.Marker, out var descriptor))
+            {
+                throw new InvalidOperationException("Failed to deserialize a component descriptor when adding a new root component.");
+            }
+
+            _ = AddComponentAsync(descriptor.ComponentType, descriptor.Parameters, selectorId.ToString(CultureInfo.InvariantCulture));
+        }
+
+        void UpdateRootComponent(RootComponentOperation<ServerComponentMarker> operation)
+        {
+            if (operation.ComponentId is not { } componentId)
+            {
+                Log.InvalidRootComponentOperation(_logger, operation.Type, message: "Missing component ID.");
+                return;
+            }
+
+            var componentState = GetComponentState(componentId);
+
+            if (!_serverComponentDeserializer.TryDeserializeSingleComponentDescriptor(operation.Marker, out var descriptor))
+            {
+                throw new InvalidOperationException("Failed to deserialize a component descriptor when updating an existing root component.");
+            }
+
+            if (descriptor.ComponentType != componentState.Component.GetType())
+            {
+                Log.InvalidRootComponentOperation(_logger, operation.Type, message: "Component type mismatch.");
+                return;
+            }
+
+            _ = RenderRootComponentAsync(componentId, descriptor.Parameters);
```

Comment:
Just to be sure, we've definitely been through DataProtection checks before we get here, right? We definitely need not to let clients supply arbitrary parameters or arbitrary component IDs.

Reply:
Yes, that happens in the call to `TryDeserializeSingleComponentDescriptor`.